File: src/commands/repo.rs
use crate::utils::{pretty_dates, request, write_to_file};
use serde_json::Value;
use std::path::PathBuf;

pub fn repo_command(owner: String, repo: String, output: Option<PathBuf>, display: bool) {
    let url = format!("https://api.github.com/repos/{}/{}", owner, repo);

    let json = request(url).expect("Failed to request data");

    if &json["message"] == "Not Found" {
        println!("Repository not found.");
        std::process::exit(0)
    }

    if !display {
        println!("{:#?}", json);
    } else {
        simplify_and_display_json(&json)
    }

    match output {
        Some(path) => {
            let result = write_to_file(serde_json::to_string_pretty(&json).unwrap(), path);
            match result {
                Ok(_) => {}
                Err(err) => println!("{}", err),
            }
        }
        None => {}
    }
}

fn simplify_and_display_json(json: &Value) {
    // General Info
    let name = json["name"].as_str().unwrap_or("None");
    let description = json["description"].as_str().unwrap_or("None");
    // Repo Status
    let forking = json["allow_forking"].as_bool().unwrap();
    let archived = json["archived"].as_bool().unwrap();
    let fork = json["fork"].as_bool().unwrap();
    let disabled = json["disabled"].as_bool().unwrap();
    let private = json["private"].as_bool().unwrap();
    // Activity Metrics
    let stars = json["stargazers_count"].as_i64().unwrap_or(0);
    let forks = json["forks_count"].as_i64().unwrap_or(0);
    let watches = json["watchers_count"].as_i64().unwrap_or(0);
    let open_issues = json["open_issues_count"].as_i64().unwrap_or(0);
    // Repository URLs
    let html_url = json["html_url"].as_str().unwrap_or("None");
    let clone_url = json["clone_url"].as_str().unwrap_or("None");
    let homepage = json["homepage"].as_str().unwrap_or("None");
    // Development Details
    let language = json["language"].as_str().unwrap_or("None");
    let default_branch = json["default_branch"].as_str().unwrap_or("None");
    let license = json["license"]["name"].as_str().unwrap_or("None");
    // Important Dates
    let created_at = json["created_at"].as_str().unwrap_or("None");
    let updated_at = json["updated_at"].as_str().unwrap_or("None");
    let pushed_at = json["pushed_at"].as_str().unwrap_or("None");

    println!("GENERAL INFO");
    println!("    {:<14}: {}", "Name", name);
    println!("    {:<14}: {}", "Description", description);
    println!();
    println!("REPO STATUS");
    println!("    {:<14}: {}", "Allows forks", forking);
    println!("    {:<14}: {}", "Archived", archived);
    println!("    {:<14}: {}", "Fork", fork);
    println!("    {:<14}: {}", "Disabled", disabled);
    println!("    {:<14}: {}", "Private", private);
    println!();
    println!("ACTIVITY METRICS");
    println!("    {:<14}: {}", "Stars", stars);
    println!("    {:<14}: {}", "Forks", forks);
    println!("    {:<14}: {}", "Watches", watches);
    println!("    {:<14}: {}", "Open issues", open_issues);
    println!();
    println!("REPOSITORY URLS");
    println!("    {:<14}: {}", "HTML URL", html_url);
    println!("    {:<14}: {}", "Clone URL", clone_url);
    if homepage != "" && homepage != "None" {
        println!("    {:<14}: {}", "Homepage", homepage);
    }
    println!();
    println!("DEVELOPMENT DETAILS");
    println!("    {:<14}: {}", "Language", language);
    println!("    {:<14}: {}", "Default branch", default_branch);
    if license != "None" {
        println!("    {:<14}: {}", "License", license);
    }
    println!();
    println!("IMPORTANT DATES");
    println!("    {:<14}: {}", "Created at", pretty_dates(created_at));
    println!("    {:<14}: {}", "Updated at", pretty_dates(updated_at));
    println!("    {:<14}: {}", "Pushed at", pretty_dates(pushed_at));
    println!();
}

File: commands/repo.go
package commands

import (
	"dev/cqb13/gstats/utils"
	"dev/cqb13/gstats/utils/ansi"
	"encoding/json"
	"fmt"
	"strings"
)

type repository struct {
	Description   string   `json:"description"`
	Fork          bool     `json:"fork"`
	CreatedAt     string   `json:"created_at"`
	UpdatedAt     string   `json:"updated_at"`
	PushedAt      string   `json:"pushed_at"`
	Homepage      string   `json:"homepage"`
	Stars         int      `json:"stargazers_count"`
	Language      string   `json:"language"`
	Forks         int      `json:"forks_count"`
	Archived      bool     `json:"archived"`
	OpenIssues    int      `json:"open_issues_count"`
	IsTemplate    bool     `json:"is_template"`
	Topics        []string `json:"topics"`
	DefaultBranch string   `json:"default_branch"`
}

func HandleRepoCommand(user string, repo string, verbose bool) {
	url := fmt.Sprintf("https://api.github.com/repos/%s/%s", user, repo)

	resp, err := utils.Get(url)
	if err != nil {
		fmt.Println(err)
		return
	}

	var rd repository

	if strings.Contains(string(resp), `"message":"Not Found"`) {
		fmt.Println("Failed to find repository")
		return
	}

	err = json.Unmarshal(resp, &rd)
	if err != nil {
		fmt.Println(err)
		return
	}

	fmt.Printf("%s%s%s%s\n", ansi.Bold, ansi.Underline, repo, ansi.Reset)
	fmt.Println(rd.Description)
	if verbose && rd.Homepage != "" {
		fmt.Printf("%s%s%s\n", ansi.Dim, rd.Homepage, ansi.Reset)
	}
	fmt.Println()
	createdAt, err := utils.RFC3339StrToPrettyStr(rd.CreatedAt)
	if err != nil {
		fmt.Println(err)
		return
	}
	pushedAt, err := utils.RFC3339StrToPrettyStr(rd.PushedAt)
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Printf("Created On: %s%-12s%sLast Push: %s%-12s%s", ansi.Bold, createdAt, ansi.Reset, ansi.Bold, pushedAt, ansi.Reset)
	if verbose {
		updatedAt, err := utils.RFC3339StrToPrettyStr(rd.UpdatedAt)
		if err != nil {
			fmt.Println(err)
			return
		}
		fmt.Printf("Last Update: %s%-12s%s\n", ansi.Bold, updatedAt, ansi.Reset)
	} else {
		fmt.Printf("\n")
	}
	fmt.Printf("Stars: %s%-10d%s", ansi.Bold, rd.Stars, ansi.Reset)
	if verbose {
		fmt.Printf("Forks: %s%-10d%sOpen Issues: %s%d%s\n", ansi.Bold, rd.Forks, ansi.Reset, ansi.Bold, rd.OpenIssues, ansi.Reset)
	} else {
		fmt.Printf("\n")
	}
	fmt.Printf("Default Branch: %s%s%-5sMain Language: %s%s%s\n", ansi.Bold, rd.DefaultBranch, ansi.Reset, ansi.Bold, rd.Language, ansi.Reset)
	fmt.Println()
	fmt.Printf("Fork: %s", ansi.Bold)
	if rd.Fork {
		fmt.Printf("True\n")
	} else {
		fmt.Printf("False\n")
	}
	if verbose {
		fmt.Printf("%sArchived: %s", ansi.Reset, ansi.Bold)
		if rd.Archived {
			fmt.Printf("True\n")
		} else {
			fmt.Printf("False\n")
		}
		fmt.Printf("%sTemplate: %s", ansi.Reset, ansi.Bold)
		if rd.IsTemplate {
			fmt.Printf("True\n")
		} else {
			fmt.Printf("False\n")
		}
	}
	fmt.Printf("%s", ansi.Reset)

	if verbose {
		fmt.Println()
		fmt.Println(createTopicStr(rd.Topics))
	}
}

func createTopicStr(topics []string) string {
	topicStr := ""
	perLine := 3

	length := len(topics) - 1

	for i, topic := range topics {
		topicStr += fmt.Sprintf("%s%s%s", ansi.Bold, topic, ansi.Reset)
		if i != length {
			topicStr += ", "
			if i%perLine == 0 && i >= perLine {
				topicStr += "\n"
			}
		}
	}

	return topicStr
}
